monitoring/interpod.go

```diff
@@ -229,28 +229,50 @@ func waitTimeoutForPodRunningInNamespace(client *kube.Clientset, podName string,
 
 // waitForPodCondition waits until a pod is in the given condition within the specified amount of time.
 func waitForPodCondition(client *kube.Clientset, ns, podName, desc string, timeout time.Duration, condition podCondition) error {
-	log.Infof("waiting up to %v for pod %s status to be %s", timeout, podName, desc)
-	for start := time.Now(); time.Since(start) < timeout; time.Sleep(pollInterval) {
-		pod, err := client.CoreV1().Pods(ns).Get(podName, metav1.GetOptions{})
-		if err != nil {
-			log.Infof("get pod %s in namespace '%s' failed, ignoring for %v: %v",
-				podName, ns, pollInterval, err)
-			continue
-		}
-		done, err := condition(pod)
-		if done {
-			// TODO: update to latest trace to wrap nil
+	log := log.WithFields(log.Fields{
+		"pod":       podName,
+		"namespace": ns,
+	})
+
+	log.Infof("waiting up to %v for pod status to be %s", timeout, desc)
+
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+
+	ticker := time.NewTicker(pollInterval)
```

Comment:
Just realizing now, that after initialization of a ticker, nothing is received on the channel until after the specified time duration. 

In this case it is probably not a big deal since `pollInterval` is only 2 seconds, so the function will not wait too long before initial execution of the embedded code block.

In places where a longer interval is used, the embedded code block might end up waiting quite a bit before initial execution.

I think we use this pattern quite often when implementing retry or a continuous loop. Might be areas for optimization.

Reply:
Yea, most areas this pattern is used does overlook the first run. In most cases it doesn't make a huge difference since it doesn't matter too much if things take a few extra seconds, but is good to keep in mind.